modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -89,86 +157,468 @@ public IgniteStatisticsRepository statisticsRepository() {
+                    objectStatistics);
+        }
+        status.doneFut.onDone();
+        */
+
+    }
+
     /**
      * Filter columns by specified names.
      *
-     * @param columns Columns to filter.
+     * @param cols Columns to filter.
      * @param colNames Column names.
      * @return Column with specified names.
      */
-    private Column[] filterColumns(Column[] columns, String... colNames) {
+    private Column[] filterColumns(Column[] cols, @Nullable Collection<String> colNames) {
         if (F.isEmpty(colNames))
-            return columns;
+            return cols;
 
         Set<String> colNamesSet = new HashSet(Arrays.asList(colNames));
-        List<Column> resultList = new ArrayList<>(colNames.length);
+        List<Column> resList = new ArrayList<>(colNames.size());
 
-        for (Column col : columns)
+        for (Column col : cols)
             if (colNamesSet.contains(col.getName()))
-                resultList.add(col);
+                resList.add(col);
 
-        return resultList.toArray(new Column[resultList.size()]);
+        return resList.toArray(new Column[resList.size()]);
     }
 
     /** {@inheritDoc} */
-    @Override public void collectObjectStatistics(String schemaName, String objName, String... colNames)
-            throws IgniteCheckedException {
-        GridH2Table tbl = schemaMgr.dataTable(schemaName, objName);
-        if (tbl == null)
-            throw new IllegalArgumentException(String.format("Can't find table %s.%s", schemaName, objName));
+    @Override public void collectObjectStatistics(
+            String schemaName,
+            String objName,
+            String... colNames
+    ) throws IgniteCheckedException {
+        StatsCollectionFutureAdapter doneFut = new StatsCollectionFutureAdapter();
+        StatsKeyMessage keyMsg = new StatsKeyMessage(schemaName, objName, Arrays.asList(colNames));
+        UUID colId = UUID.randomUUID();
+        StatCollectionStatus status = new StatCollectionStatus(colId, Collections.singletonList(keyMsg),
+            Collections.emptyMap(), doneFut);
+        synchronized (status) {
+            currCollections.updateCollection(colId, s -> status);
+            Map<StatsKeyMessage, int[]> failedPartitions = null;
+            int cnt = 0;
+            do {
+                Collection<StatsCollectionAddrRequest> reqs = currCollections.generateCollectionRequests(colId,
+                    Collections.singletonList(keyMsg), failedPartitions);
+                Map<UUID, StatsCollectionRequest> msgs = reqs.stream().collect(Collectors.toMap(
+                    StatsCollectionAddrRequest::nodeId, StatsCollectionAddrRequest::req));
 
-        if (log.isDebugEnabled())
-            log.debug(String.format("Starting statistics collection by %s.%s object", schemaName, objName));
+                Map<UUID, StatsCollectionRequest> failedMsgs = sendLocalRequests(msgs);
```

Comment:
What does 'sendLocalRequests' mean? Why 'local'?

Reply:
I thought about locallity of requests. All of them should either collect or cancel local operations. But let it be just sendRequests.